### ISSUE TITLE
Are vec swizzles from vec<T, N> to vec<T, N+n> allowed?

### ISSUE THREAD

**Issue by illuhad** [KhronosGroup/SYCL-Docs#521]
It recently was reported by a user that swizzling to a larger `vec` size currently does not work in AdaptiveCpp, while it seems to work in DPC++.
Here is an example of the code in question:

```c++
#define SYCL_SIMPLE_SWIZZLES
#include <sycl/sycl.hpp>

int main(int argc, char* argv[]) {
    sycl::float2 p0 {1.0f, 2.0f};
    sycl::float3 p1 = p0.xyx(); // Error occurs here
    return 0;
}
```

The specification says:
> Where XYZW_SWIZZLE is all permutations with repetition, of any subset with length greater than 1, of x, y for NumElements == 2, x, y, z for NumElements == 3 and x, y, z, w for NumElements == 4. For example a four element vec provides permutations including xzyw, xyyy and xz.

The way I read it, the 3-component `xyx` in the example is not a permutation of any subset of `(x,y)` (since `NumElements == 2`) and thus would not be allowed. I'm not sure if the exact mathematical definition of "permutation with repetition" allows more freedom here, but in that case it should perhaps be made more obvious. The current situation is already that AdaptiveCpp and DPC++ implement this differently.

This is the opposite case of https://github.com/KhronosGroup/SYCL-Docs/issues/29

**Comment by Pennycook** [comment]
I agree that this needs to be clarified.  I originally thought that because `xyyy` was listed in the part of the specification you quoted, `xyx` would also work... But I see now how this could be read as `xyyy` only being permitted for vectors which already have four elements.

FWIW, I think this ambiguity is something that we inherited from OpenCL.  I can't see anything in [Section 6.3.7](https://registry.khronos.org/OpenCL/specs/3.0-unified/html/OpenCL_C.html#vector-components) of the OpenCL C specification that explicitly talks about this case.  There's one example where the number of components is larger than the size of the vector:

```opencl
float4 a, b, c, d;

// illegal - component a.xxxxxxx is not a valid vector type
x = (float16)(a.xxxxxxx, b.xyz, c.xyz, d.xyz);
```

...and my interpretation of this is that it's illegal because there are 7 `x`s, not because there are more than 4 `x`s.

I can't see a conformance test for this sort of thing in the [OpenCL-CTS](https://github.com/KhronosGroup/OpenCL-CTS/blob/b36b2d592c7c51b76be8950f1c2a7dc6184c2064/test_conformance/basic/test_vector_swizzle.cpp) either.

The definition in [SPIR-V](https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#OpVectorShuffle) seems clearer, and by my reading says:
- The number of components in the result doesn't have to be the same as the number of components in the vector (i.e., it can be smaller _or_ larger).
- Components can be selected in any order.
- Components can be repeated.

Since the SPIR-V behavior is clearest, my vote would be to follow that, and introduce text to align SYCL and OpenCL with the SPIR-V definition. But I'm interested to hear what @bashbaug, since my reading of the OpenCL and SPIR-V specifications may be wrong here.

**Comment by gmlueck** [comment]
Is there also disagreement about the templated member function `vec::swizzle`?  For example, does AdaptiveCpp allow something like this:

```
#include <sycl/sycl.hpp>

int main(int argc, char* argv[]) {
    sycl::float2 p0 {1.0f, 2.0f};
    sycl::float3 p1 = p0.swizzle<0,1,0>();
    return 0;
}
```

**Comment by illuhad** [comment]
@gmlueck I expect that your example will work in AdaptiveCpp (without having explicitly tested it).

I think the problem is only in the wording for the simple swizzles. Simple swizzles are only allowed for a subset of possible swizzles that can be expressed with the `swizzle()` member function, and it seems that the precise definition of that subset is currently somewhat ambiguous.
